stablehlo/dialect/AssemblyFormat.cpp

```diff
@@ -300,6 +301,38 @@ ParseResult parseDenseI64Array(OpAsmParser& parser,
   return success();
 }
 
+void printDynamicShape(AsmPrinter& p, llvm::ArrayRef<int64_t> shape) {
+  auto printIntOrQuestion = [&](int64_t value) {
+    if (ShapedType::isDynamic(value))
+      p << '?';
+    else
+      p << value;
+  };
+  p << " [";
```

Comment:
Why does this function print a leading whitespace and the functions below don't?

Reply:
I'll add this to my list to investigate. This may be an upstream bug. There is no space automatically added between literals and custom directives when specifying declarative assembly _for attributes_. There is a space for ops.

For now, I'll keep the custom declarations consistent and add an explicit space to the tablegen declaration.